src/runner.rs

```diff
@@ -1454,3 +1448,97 @@ impl ValueStack {
         self.sp
     }
 }
+
+struct CallStack {
+    buf: Vec<FunctionContext>,
+    limit: usize,
+}
+
+impl CallStack {
+    fn push(&mut self, ctx: FunctionContext) {
+        self.buf.push(ctx);
+    }
+
+    fn pop(&mut self) -> Option<FunctionContext> {
+        self.buf.pop()
+    }
+
+    fn is_empty(&self) -> bool {
+        self.buf.is_empty()
+    }
+
+    fn is_full(&self) -> bool {
+        self.buf.len() + 1 >= self.limit
+    }
+}
+
+/// Used to recycle stacks instead of allocating them repeatedly.
+pub struct StackRecycler {
+    value_stack_buf: Option<Box<[RuntimeValueInternal]>>,
+    value_stack_limit: usize,
+    call_stack_buf: Option<Vec<FunctionContext>>,
+    call_stack_limit: usize,
+}
+
+impl StackRecycler {
+    /// Limit stacks created by this recycler to
+    /// - `value_stack_limit` bytes for values and
+    /// - `call_stack_limit` levels for calls.
+    pub fn with_limits(value_stack_limit: usize, call_stack_limit: usize) -> Self {
+        Self {
+            value_stack_buf: None,
+            value_stack_limit,
+            call_stack_buf: None,
+            call_stack_limit,
+        }
+    }
+
+    fn recreate_value_stack(this: &mut Option<&mut Self>) -> ValueStack {
+        let limit = this
+            .as_ref()
+            .map_or(DEFAULT_VALUE_STACK_LIMIT, |this| this.value_stack_limit)
+            / ::core::mem::size_of::<RuntimeValueInternal>();
```

Comment:
This used to be `size_of::<RuntimeValue>` but I suspect this was a typo as `RuntimeValue` does include the tag in addition to the 64 bits of payload.